// @sap/cds/libx/odata/middleware/batch.js
const cds = require('../../../')

const express = require('express')
const { STATUS_CODES } = require('http')
const qs = require('querystring')
const { URL } = require('url')

const multipartToJson = require('../parse/multipartToJson')
const { getBoundary } = require('../utils')

const { odataError } = require('./error')

const HTTP_METHODS = { GET: 1, POST: 1, PUT: 1, PATCH: 1, DELETE: 1 }
const CT = { JSON: 'application/json', MULTIPART: 'multipart/mixed' }
const CRLF = '\r\n'

/*
 * deconstruct and validate
 */

const _deserializationError = message => cds.error({ status: 400, message: `Deserialization Error: ${message}` })

// Function must be called with an object containing exactly one key-value pair representing the property name and its value
const _validateProperty = (name, value, type) => {
  if (value === undefined) throw _deserializationError(`Parameter '${name}' must not be undefined.`)

  switch (type) {
    case 'Array':
      if (!Array.isArray(value)) throw _deserializationError(`Parameter '${name}' must be type of '${type}'.`)
      break
    default:
      if (typeof value !== type) throw _deserializationError(`Parameter '${name}' must be type of '${type}'.`)
  }
}

const _deconstructAndValidateBatchBody = body => {
  const { requests } = body

  _validateProperty('requests', requests, 'Array')
  if (!requests.length) cds.error({ status: 400, message: 'There must be at least one request' })
  if (requests.length > cds.env.odata.batch_limit) cds.error({ status: 429, message: 'BATCH_TOO_MANY_REQ' })

  const ids = {}
  const atomicityGroups = []

  let previousAtomicityGroup

  for (let i = 0; i < requests.length; i++) {
    const request = requests[i]

    if (typeof request !== 'object')
      throw _deserializationError(`Element of 'requests' array at index ${i} must be type of 'object'.`)

    const { id, method, url, body, atomicityGroup } = request

    if (atomicityGroup && atomicityGroups.length && atomicityGroup === atomicityGroups.at(-1)[0]?.atomicityGroup)
      atomicityGroups.at(-1).push(request)
    else atomicityGroups.push([request])

    _validateProperty('id', id, 'string')

    if (ids[id]) throw _deserializationError(`Request ID '${id}' is not unique.`)
    else ids[id] = request

    // TODO: validate allowed methods or let express throw the error?
    _validateProperty('method', method, 'string')
    if (!(method.toUpperCase() in HTTP_METHODS))
      throw _deserializationError(`Method '${method}' is not allowed. Only DELETE, GET, PATCH, POST or PUT are.`)

    _validateProperty('url', url, 'string')
    if (url.startsWith('/$batch')) throw _deserializationError('Nested batch requests are not allowed.')

    // TODO: support for non JSON bodies?
    if (body !== undefined && typeof body !== 'object')
      throw _deserializationError('A Content-Type header has to be specified for a non JSON body.')

    // TODO
    // if (!(method.toUpperCase() in { GET: 1, DELETE: 1 }) && !body)
    //   throw _deserializationError(`Body is required for ${method} requests.`)

    if (atomicityGroup) {
      _validateProperty('atomicityGroup', atomicityGroup, 'string')

      // All request objects with the same value for atomicityGroup MUST be adjacent in the requests array
      if (atomicityGroup !== previousAtomicityGroup) {
        if (ids[atomicityGroup]) throw _deserializationError(`Atomicity group ID '${atomicityGroup}' is not unique.`)
        else ids[atomicityGroup] = []
      }
      ids[atomicityGroup].push(request)
    }

    if (url.startsWith('$') && url !== '$metadata') {
      request.dependsOn ??= []
      const dependencyId = url.split('/')[0].replace(/^\$/, '')
      if (!request.dependsOn.includes(dependencyId)) {
        request.dependsOn.push(dependencyId)
      }
    }

    if (request.dependsOn) {
      _validateProperty('dependsOn', request.dependsOn, 'Array')
      request.dependsOn.forEach(dependsOnId => {
        _validateProperty('dependent request ID', dependsOnId, 'string')

        const dependency = ids[dependsOnId]
        if (!dependency) {
          throw _deserializationError(
            `"${dependsOnId}" does not match the id or atomicity group of any preceding request`
          )
        }

        // automatically add the atomicityGroup of the dependency as a dependency (actually a client error)
        const dag = dependency.atomicityGroup
        if (dag && dag !== atomicityGroup && !request.dependsOn.includes(dag)) {
          request.dependsOn.push(dag)
        }
      })
    }

    // TODO: validate if, and headers

    previousAtomicityGroup = atomicityGroup
  }

  return { ids, atomicityGroups }
}

/*
 * subrequest (a.k.a. lookalike)
 */

let error_mws

// REVISIT: Why not simply use {__proto__:req, ...}?
const _createSubrequest = (request, _req, _res) => {
  error_mws ??= cds.middlewares.after.filter(mw => mw.length === 4) // error middleware has 4 params

  const subrequest = { id: request.id }

  // req
  const req = (subrequest.req = new express.request.constructor())
  req.__proto__ = express.request
  req.app = _req.app
  req.method = request.method.toUpperCase()
  req.url = request.url
  const u = new URL(request.url, 'http://cap')
  req.query = qs.parse(u.search.slice(1))
  req.headers = request.headers || {}
  req.headers['content-type'] ??= _req.headers['content-type']
  if (request.content_id) req.headers['content-id'] = request.content_id
  req.body = request.body
  if (_req._login) req._login = _req._login
  // REVISIT: mark as subrequest (only needed for logging)
  req._subrequest = true

  // res
  const res = (req.res = subrequest.res = new express.response.constructor(req))
  res.__proto__ = express.response
  res.app = _res.app

  // next
  let _err
  subrequest.next = err => {
    _err = err
    let _next_called
    const _next = err => {
      _err = err
      _next_called = true
    }
    for (const mw of error_mws) {
      _next_called = false
      mw(_err, req, res, _next)
      if (!_next_called) break //> next chain was interrupted -> done
    }
    if (_next_called) {
      // here, final error middleware called next (which actually shouldn't happen!)
      if (_err.statusCode) res.status(_err.statusCode)
      if (typeof _err === 'object') res.json({ error: _err })
      else res.send(_err)
    }
  }

  // resolve/reject promise for subrequest via res.end()
  subrequest.promise = new Promise((resolve, reject) => {
    res.json = function (obj) {
      subrequest._json = obj
      return this.__proto__.json.call(this, obj)
    }
    res.end = (chunk, encoding) => {
      res._chunk = chunk
      res._encoding = encoding
      if (res.statusCode >= 400) return reject(_err)
      resolve(subrequest)
    }
  })

  return subrequest
}

/*
 * multipart/mixed response
 */

const _tryParse = body => {
  try {
    return JSON.parse(body)
  } catch {
    return body
  }
}

const _formatResponseMultipart = response => {
  const { content_id, statusCode, headers } = response
  let { body } = response

  let txt = `content-type: application/http${CRLF}content-transfer-encoding: binary${CRLF}`
  if (content_id) txt += `content-id: ${content_id}${CRLF}`
  txt += CRLF
  txt += `HTTP/1.1 ${statusCode} ${STATUS_CODES[statusCode]}${CRLF}`

  for (const key in headers) txt += key + ': ' + headers[key] + CRLF
  txt += CRLF

  if (body) {
    if (Buffer.isBuffer(body)) body = body.toString()
    body = _tryParse(body)
    if (body && typeof body === 'object') {
      let meta = [],
        data = []
      for (const [k, v] of Object.entries(body)) {
        if (k.startsWith('@')) meta.push(`"${k}":${typeof v === 'string' ? `"${v.replaceAll('"', '\\"')}"` : v}`)
        else data.push(JSON.stringify({ [k]: v }).slice(1, -1))
      }
      const _json_as_txt = '{' + meta.join(',') + (meta.length && data.length ? ',' : '') + data.join(',') + '}'
      body = _json_as_txt
    }
    txt += body
  }

  return [txt]
}

const _writeResponseMultipart = (responses, res, boundary) => {
  res.write(`--${boundary}${CRLF}`)

  const rejected = responses.find(r => r.status === 'fail')
  if (rejected) {
    res.write(`${_formatResponseMultipart(rejected)[0]}${CRLF}`)
    return
  }

  const groupId = responses[0].atomicityGroup
  if (groupId) res.write(`content-type: multipart/mixed;boundary=${groupId}${CRLF}${CRLF}`)

  for (const response of responses) {
    for (const each of _formatResponseMultipart(response)) {
      if (groupId) res.write(`--${groupId}${CRLF}`)
      res.write(`${each}${CRLF}`)
    }
  }

  if (groupId) res.write(`--${groupId}--${CRLF}`)
}

/*
 * application/json response
 */

const _formatStatics = {
  comma: ','.charCodeAt(0),
  body: Buffer.from('"body":'),
  close: Buffer.from('}')
}

const _formatResponseJson = response => {
  const { id, atomicityGroup, statusCode, headers, body } = response

  const chunk = { id, status: statusCode, headers }
  if (atomicityGroup) chunk.atomicityGroup = atomicityGroup

  const raw = Buffer.from(JSON.stringify(chunk))

  // body?
  if (!body) return [raw]

  // change last "}" into ","
  raw[raw.byteLength - 1] = _formatStatics.comma

  return [
    raw,
    _formatStatics.body,
    // Stringify non-JSON bodies
    !chunk.headers?.['content-type']?.includes('application/json')
      ? JSON.stringify(Buffer.isBuffer(body) ? body.toString() : body)
      : body,
    _formatStatics.close
  ]
}

const _writeResponseJson = (responses, res) => {
  for (const response of responses) {
    if (res._separator) res.write(res._separator)
    else res._separator = Buffer.from(',')

    for (const each of _formatResponseJson(response)) res.write(each)
  }
}

/*
 * helpers
 */

const _urlWithResolvedDependency = (dependsOnId, results, request, req, srv) => {
  const dependentResult = results.find(r => r.value.id === dependsOnId)
  const dependentOnUrl = dependentResult.value.req.originalUrl
  const dependentOnResult = dependentResult.value._json ?? JSON.parse(dependentResult.value.res._chunk)
  const recentUrl = request.url
  const cqn = cds.odata.parse(dependentOnUrl, { service: srv, baseUrl: req.baseUrl, strict: true })
  const target = cds.infer.target(cqn)
  const keyString =
    '(' +
    [...target.keys]
      .filter(k => !k.isAssociation)
      .map(k => {
        let v = dependentOnResult[k.name]
        if (typeof v === 'string' && k._type !== 'cds.UUID') v = `'${v}'`
        return k.name + '=' + v
      })
      .join(',') +
    ')'
  return recentUrl.replace(`$${dependsOnId}`, dependentOnUrl + keyString)
}

const _resolveDependencies = async (request, deps, responses, ids, req, srv) => {
  const { url } = request

  const results = await Promise.allSettled(deps.map(d => d.promise))

  const failed = results.some(({ status }) => status === 'rejected') || deps.some(d => d._rejected_on_commit)
  if (failed) {
    const err = { code: '424', message: 'Failed Dependency' }
    responses[request._index] = _getResponse(request, { res: { statusCode: 424 } }, err)
    throw err
  }

  const dependsOnId = url.split('/')[0].replace(/^\$/, '')
  if (dependsOnId in ids) request.url = _urlWithResolvedDependency(dependsOnId, results, request, req, srv)
}

const _getResponse = (request, { res }, error) => {
  const { id, content_id, atomicityGroup, _index } = request
  return {
    id,
    content_id,
    atomicityGroup,
    _index,
    status: error ? 'fail' : 'ok',
    statusCode: res.statusCode,
    body: error ? { error } : res._chunk,
    headers: res.getHeaders?.() ?? {}
  }
}

// construct odata-compliant error (without modifying original error)
const _getODataError = err => {
  return odataError(Object.create(err), {
    get locale() {
      return cds.context.locale
    },
    get() {}
  })
}

const _replaceResponsesWithCommitErrors = (err, responses, ids) => {
  const error = _getODataError(err)

  // adjust all responses to commit error and mark respective promises as failed
  const body = JSON.stringify({ error })
  const length = Buffer.byteLength(body, 'utf8')
  for (const response of responses) {
    response.status = 'fail'
    response.statusCode = error.status
    response.body = body
    response.headers ??= {}
    response.headers['content-length'] = length

    ids[response.id]._rejected_on_commit = true
    if (response.atomicityGroup) ids[response.atomicityGroup]._rejected_on_commit = true
  }
}

const _serializeErrors = responses => {
  for (const response of responses) {
    if (response.status === 'fail' && typeof response.body === 'object') {
      if (response.body.error && !response.body.error.toJSON) response.body.error = _getODataError(response.body.error)
      response.body = JSON.stringify(response.body)
    }
  }
}

/*
 * process
 */

const _txs = {}
cds.once('shutdown', () => Promise.all(Object.values(_txs).map(tx => tx.rollback().catch(() => {}))))

const _processBatch = async (srv, router, req, res, next, body, ct, boundary) => {
  body ??= req.body
  ct ??= 'JSON'
  // respond with requested content type (i.e., accept) with fallback to the content type used in the request
  let isJson = ct === 'JSON'
  if (req.headers.accept) {
    if (req.headers.accept.indexOf('multipart/mixed') > -1) isJson = false
    else if (req.headers.accept.indexOf('application/json') > -1) isJson = true
  }

  // continue-on-error defaults to true in json batch
  let continue_on_error = req.headers.prefer?.match(/odata\.continue-on-error(=(\w+))?/)
  if (!continue_on_error) {
    continue_on_error = isJson ? true : false
  } else {
    continue_on_error = continue_on_error[2] === 'false' ? false : true
  }

  try {
    let { ids, atomicityGroups } = _deconstructAndValidateBatchBody(body)

    const only_gets = body.requests.every(r => r.method === 'GET')
    const max_parallel = only_gets ? cds.env.odata.max_batch_parallelization : 1

    // experimental!
    const _only_individual_requests = () =>
      atomicityGroups.every(ag => ag.length === 1) && body.requests.every(r => !r.dependsOn?.length)
    if (only_gets && cds.env.odata.group_parallel_gets && _only_individual_requests())
      atomicityGroups = [atomicityGroups.reduce((acc, cur) => (acc.push(...cur), acc), [])]

    res.setHeader('Content-Type', isJson ? CT.JSON : CT.MULTIPART + ';boundary=' + boundary)
    res.status(200)
    res.write(isJson ? '{"responses":[' : '')

    const queue = []
    let _continue = true
    const _queued_exec = (atomicityGroup, agIndex, responses = []) => {
      const groupId = atomicityGroup[0].atomicityGroup

      return new Promise(resolve => queue.push(resolve)).then(() => {
        if (!_continue) return

        const promise = srv
          .tx(tx => {
            // ensure rollback on shutdown so the db can disconnect
            const _id = cds.utils.uuid()
            _txs[_id] = tx
            tx.context.on('done', () => delete _txs[_id])

            const subrequests = []
            for (let i = 0; i < atomicityGroup.length; i++) {
              const request = atomicityGroup[i]

              // for multipart, we need to keep the original order of requests in the atomicity group
              request._index = i

              const { id, dependsOn } = request

              // wait for/ resolve dependencies?
              const deps = dependsOn?.filter(id => id !== groupId).map(id => ids[id])
              if (deps) ids[id].promise = _resolveDependencies(request, deps, responses, ids, req, srv)
              else ids[id].promise = Promise.resolve()

              ids[id].promise = ids[id].promise.then(() => {
                const subrequest = _createSubrequest(request, req, res)
                router.handle(subrequest.req, subrequest.res, subrequest.next)
                return subrequest.promise
                  .then(res => {
                    responses[request._index] = _getResponse(request, subrequest)
                    return res
                  })
                  .catch(err => {
                    responses[request._index] = _getResponse(request, subrequest, err)
                    throw err
                  })
              })

              subrequests.push(ids[id].promise)
            }

            // ensure all subrequests run in this tx
            // (if first subrequest fails without really opening the tx, the rest are executed in a "dangling tx")
            return Promise.allSettled(subrequests)
              .then(ress => {
                // wait for all previous atomicity groups (ignoring errors via allSettled) for odata v2
                const prevs = []
                for (let i = 0; i < agIndex; i++) prevs.push(promises[i])
                return Promise.allSettled(prevs).then(() => ress)
              })
              .then(ress => {
                const failed = ress.filter(({ status }) => status === 'rejected')
                if (!failed.length) return
                // throw first error and call srv.on('error') for the others
                const first = failed.shift()
                if (srv.handlers._error?.length)
                  for (const other of failed)
                    for (const each of srv.handlers._error) each.handler.call(srv, other.reason, cds.context)
                throw first.reason
              })
          })
          .catch(err => {
            responses._has_failure = true

            // abort batch on first failure with odata.continue-on-error: false
            if (!continue_on_error) _continue = false

            if (!responses.some(r => r.status === 'fail')) {
              // here, the commit was rejected even though all requests were successful (e.g., by custom handler or db consistency check)
              _replaceResponsesWithCommitErrors(err, responses, ids)
            }
          })
          .finally(() => {
            // trigger next in queue
            if (queue.length) queue.shift()()

            // late error serialization
            if (responses._has_failure) _serializeErrors(responses)

            if (isJson) _writeResponseJson(responses, res)
            else _writeResponseMultipart(responses, res, boundary)
          })

        if (ids[groupId]) ids[groupId].promise = promise

        return promise
      })
    }

    // queue execution of atomicity groups
    const promises = []
    for (let i = 0; i < atomicityGroups.length; i++) promises.push(_queued_exec(atomicityGroups[i], i))

    // trigger first max_parallel in queue
    for (let i = 0; i < max_parallel; i++) if (queue.length) queue.shift()()

    await Promise.all(promises)

    res.write(isJson ? ']}' : `--${boundary}--${CRLF}`)
    res.end()
  } catch (e) {
    next(e)
  }
}

/*
 * exports
 */

module.exports = adapter => {
  const { router, service } = adapter

  const textBodyParser = express.text({
    ...adapter.body_parser_options,
    type: '*/*' // REVISIT: why do we need to override type here?
  })

  return function odata_batch(req, res, next) {
    if (req.method !== 'POST') {
      cds.error({ status: 405, message: `Method ${req.method} is not allowed for calls to $batch endpoint` })
    }

    if (req.headers['content-type'].includes('application/json')) {
      return _processBatch(service, router, req, res, next)
    }

    if (req.headers['content-type'].includes('multipart/mixed')) {
      return textBodyParser(req, res, async function odata_batch_next(err) {
        if (err) return next(err)

        const boundary = getBoundary(req)
        if (!boundary) return next(new cds.error({ status: 400, message: 'No boundary found in Content-Type header' }))

        try {
          const { requests } = await multipartToJson(req.body, boundary)
          _processBatch(service, router, req, res, next, { requests }, 'MULTIPART', boundary)
        } catch (e) {
          // REVISIT: (how) handle multipart accepts?
          next(e)
        }
      })
    }

    cds.error({
      status: 400,
      message: 'Batch requests must have content type multipart/mixed or application/json'
    })
  }
}
